pkg/operation/botanist/etcd.go

```diff
@@ -191,6 +192,65 @@ func (b *Botanist) scaleETCD(ctx context.Context, replicas int32) error {
 	return b.Shoot.Components.ControlPlane.EtcdEvents.Scale(ctx, replicas)
 }
 
+func (b *Botanist) deployOrRestoreMainEtcd(ctx context.Context) error {
+	isRestoreRequired, err := b.isRestorationOfMultiNodeMainEtcdRequired(ctx)
+	if err != nil {
+		return err
+	}
+
+	if isRestoreRequired {
+		return b.restoreMultiNodeMainEtcd(ctx)
+	}
+
+	return b.Shoot.Components.ControlPlane.EtcdMain.Deploy(ctx)
+}
+
+func (b *Botanist) isRestorationOfMultiNodeMainEtcdRequired(ctx context.Context) (bool, error) {
+	if !b.isRestorePhase() || !v1beta1helper.IsHAControlPlaneConfigured(b.Shoot.GetInfo()) {
+		return false, nil
+	}
+
+	etcd, err := b.Shoot.Components.ControlPlane.EtcdMain.Get(ctx)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+
+	// The etcd has already been scaled up to the desired number of replicas
+	// and therefore has been restored, then it has been scaled down to 0
+	// replicas as part of the reconciliation flow for hibernated shoots.
+	if b.Shoot.HibernationEnabled && etcd.Spec.Replicas == 0 {
+		return false, nil
+	}
+	// The etcd has already been scaled up to the desired number of replicas
+	// and therefore has been restored.
+	if etcd.Spec.Replicas == getEtcdReplicas(b.Shoot.GetInfo()) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (b *Botanist) restoreMultiNodeMainEtcd(ctx context.Context) error {
+	originalReplicas := b.Shoot.Components.ControlPlane.EtcdMain.GetReplicas()
+	defer func() {
+		// Revert the original replica count for the etcd. This is done in case a step
+		// is added to the reconciliation flow that depends on the etcd's replica count.
+		b.Shoot.Components.ControlPlane.EtcdMain.SetReplicas(originalReplicas)
+	}()
+
+	b.Shoot.Components.ControlPlane.EtcdMain.SetReplicas(pointer.Int32(1))
+	if err := b.Shoot.Components.ControlPlane.EtcdMain.Deploy(ctx); err != nil {
+		return err
+	}
+	if err := b.Shoot.Components.ControlPlane.EtcdMain.Wait(ctx); err != nil {
+		return err
+	}
+	return b.Shoot.Components.ControlPlane.EtcdMain.Scale(ctx, getEtcdReplicas(b.Shoot.GetInfo()))
```

Comment:
```suggestion
	return b.Shoot.Components.ControlPlane.EtcdMain.Scale(ctx, originalReplicas)
```
I guess we can use `originalReplicas` instead of calling `getEtcdReplicas(b.Shoot.GetInfo())` right ?

Reply:
`originalReplicas` can be `nil` if the shoot is hibernated. When migrating hibernated shoots we have to do perform the same steps - deploy etcd, scale it to 1, then scale it to 3, scaling it back to 0 will be taken care of later on here: https://github.com/gardener/gardener/blob/6ac1c2d6c2357725ca0d5525c8586abe8c0c5a18/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L698-L702 
more specifically here: https://github.com/gardener/gardener/blob/7c647ba75fa5bd6bd9db590c43f91b6b764722a7/pkg/operation/botanist/controlplane.go#L150
If the same steps are not taken for hibernated clusters, when a hibernated cluster is migrated and at a later point woken up, I observed the following error: 
```
peer[7610a9b56fcea62c]=cdf818194e3a8c32, local=e808882a00e8ba84)
2023-03-15 16:37:44.227223 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[3ea9caa65f20e737]=cdf818194e3a8c32, local=e808882a00e8ba84)
2023-03-15 16:37:44.227845 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[7610a9b56fcea62c]=cdf818194e3a8c32, local=e808882a00e8ba84)
2023-03-15 16:37:44.228000 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[3ea9caa65f20e737]=cdf818194e3a8c32, local=e808882a00e8ba84)
2023-03-15 16:37:44.231458 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[7610a9b56fcea62c]=cdf818194e3a8c32, local=e808882a00e8ba84)
```

Reply:
got it, thanks